src/fqdncache.cc

```diff
@@ -304,7 +304,7 @@ fqdncacheCallback(fqdncache_entry * f, int wait)
     f->handler = nullptr;
 
     if (cbdataReferenceValidDone(f->handlerData, &cbdata)) {
-        const Dns::LookupDetails details(f->error_message, wait);
+        const Dns::LookupDetails details(SBuf(f->error_message), wait);
```

Comment:
In order to use std::optional this caller code needs to determine whether an error message exists and construct LookupDetails with different constructors - one which sets error, and one which only sets wait time (not yet existing).

Reply:
> In order to use std::optional this caller code needs to determine whether an error message exists and construct LookupDetails with different constructors

There is no such requirement. LookupDetails constructor can work correctly with the existing SBuf API. It is not the most efficient way, but it is simple and is a good _next_ step.

Said that, it is possible, perhaps even likely, that `f->error_message` should be converted to std::optional as well and/or that would we should add another constructor or make some other API changes. I am moving one step at a time here.

I will not repeat this comment in other similar contexts, but it does apply.

Reply:
AFAICS, kinkie is trying to take a single refactor step in this PR (String->SBuf only). Whereas you are requesting a skip straight to step 3, ignoring the additional changes of step 2 required to make the step 3 be as efficient as you claim it is.

Reply:
> AFAICS, kinkie is trying to take a single refactor step in this PR (String->SBuf only).

Yes, and stopping at that first step does more harm than good in this particular case. We have to spend time reviewing every `error` user anyway. We should use that to switch to the _correct_ type for this optional field instead of raw SBuf.

> Whereas you are requesting a skip straight to step 3, ignoring the additional changes of step 2 required to make the step 3 be as efficient as you claim it is.

Or you can say that I am requesting to go straight to step 25. The number of steps does not really matter here because these steps are tiny, natural, there are people willing to make them, and this PR is the right/best place to make them.

Reply:
> > AFAICS, kinkie is trying to take a single refactor step in this PR (String->SBuf only).
> 
> Yes, and stopping at that first step does more harm than good in this particular case. We have to spend time reviewing every `error` user anyway. We should use that to switch to the _correct_ type for this optional field instead of raw SBuf.

@yadij is correct, I am taking an incremental approach.
I also agree that moving to std::optional is conceptually more correct, because the current API being "empty string == no error" can be improved upon.
I do not agree that stopping at String -> SBuf is worse than the current status. I entered this specific rabbit hole because of a single caller (ipcache) needing to keep around a c-style string (the hash index) to pass it to DNS lookup which is inefficient and makes memory ownership unclear, and that is the part I would like to focus on next, as we discussed elsewhere in this PR.
So by stopping here, we would be missing an opportunity to improve an API but not make it worse; it would instead give me an opportunity to focus on an interesting use case and get benefits right away.
A way to get a bit of both is to add a boolean getter which for now would check the string length, and eventually move to checking std::optional. We could use it to clarify intent on part of the callers and facilitate the transition.

> > Whereas you are requesting a skip straight to step 3, ignoring the additional changes of step 2 required to make the step 3 be as efficient as you claim it is.
> 
> Or you can say that I am requesting to go straight to step 25. The number of steps does not really matter here because these steps are tiny, natural, there are people willing to make them, and this PR is the right/best place to make them.

I would rather use followup PRs; what do you think of my proposed strategy above?

Reply:
> stopping here, we would be missing an opportunity to improve an API but ... give me an opportunity to focus on an interesting use case and get benefits right away.

Nothing in my change requests prevents your from focusing on whatever you want to focus on or delaying benefits by any amount of time worth talking about, especially given the amount of time already spent on this PR. Even if you did not want to follow my suggestions, you could simply say "if you can do it quickly, please address your change requests and we will see whether the results are worth keeping". That takes approximately one minute for you to write and 30 minutes for me to implement. Compare that with more than an hour we have already _wasted_ on useless discussions.

> A way to get a bit of both is to add a boolean getter

Please see #1241 for what I consider the correct way to address my change requests. That PR does not add any methods.

> I would rather use followup PRs; what do you think of my proposed strategy above?

I would rather merge #1241 which starts with where this PR is right now and then addresses my change requests and Amos' change requests (that did not reach wrong conclusions based on unfinished PR code).

Followup PRs are needed indeed, but #1241 represents a minimal change IMO (not from "a minimal change that compiles" point of view, but from a "minimal change that is worth merging" point of view).

Reply:
@kinkie, please consider closing this PR because its improvements are covered by the merged PR #1241. Thank you.